### ISSUE TITLE
Convert data pipeline to use TRL PPOTrainer expected dataset format

### ISSUE THREAD

**Issue by peara** [peara/rlvr-summary#23]
## Description

Currently, our data pipeline produces raw text format (`{"article": "...", "highlights": "...", "id": "..."}`) while TRL's PPOTrainer expects tokenized prompts in HuggingFace Dataset format (`{"input_ids": [...], "attention_mask": [...]}`). We need to adapt our pipeline to TRL's expected format for optimal performance and compatibility.

## Background

**Current Pipeline Format:**
```python
{
    "id": "cnn_0001",
    "article": "Scientists have discovered...",
    "highlights": "Scientists discover new butterfly species..."
}
```

**TRL PPOTrainer Expected Format:**
```python
{
    "input_ids": [101, 2054, 2003, ...],  # Tokenized prompt only
    "attention_mask": [1, 1, 1, ...],
    "query": "Summarize: Scientists have discovered...\n\nSummary:",  # Optional
    "reference": "Scientists discover new butterfly species..."  # For reward computation
}
```

### PPOTrainer Requirements (from TRL source code analysis):
1. **Dataset Input**: Expects `Dataset` object with tokenized `input_ids` representing **prompts only** (not prompt + completion)
2. **Generation**: Trainer handles completion generation during training 
3. **Rewards**: Computed on generated text vs. reference summaries
4. **Training Loop**: Fully managed by TRL (generation → reward → PPO updates)

### Benefits of TRL Format:
- ✅ Highly optimized for memory and performance
- ✅ Built-in checkpointing, logging, evaluation
- ✅ Better HuggingFace ecosystem integration
- ✅ Less custom PPO implementation to maintain
- ✅ Automatic handling of padding, batching, device placement

### Core Changes
- [ ] **Modify `PPOTrainingLoop.load_datasets()`** to return TRL-compatible Dataset objects
- [ ] **Create `_convert_to_ppo_format()` method** to transform processed data
- [ ] **Update prompt template handling** for consistent tokenization
- [ ] **Integrate reward computation** with TRL's generation cycle
- [ ] **Remove custom training loop fallback** once TRL integration is stable

### Data Pipeline Integration
- [ ] **Keep existing data validation pipeline** (CNNDMLoader → TextPreprocessor → DataValidator)
- [ ] **Add tokenization step** after validation but before TRL format conversion
- [ ] **Handle truncation/padding** according to model requirements
- [ ] **Preserve metadata** (article text, reference summaries) for reward computation

#### 1. Update PPOTrainingLoop.load_datasets()
```python
def load_datasets(self) -> Tuple[Dataset, Dataset]:
    # Use existing pipeline for data loading/validation
    train_data, eval_data = self._load_and_process_data()
    
    # Convert to TRL format
    train_dataset = self._convert_to_ppo_format(train_data)
    eval_dataset = self._convert_to_ppo_format(eval_data)
    
    return train_dataset, eval_dataset
```

#### 2. Create TRL format converter
```python
def _convert_to_ppo_format(self, data: List[Dict]) -> Dataset:
    ppo_samples = []
    for sample in data:
        # Create standardized prompt
        prompt = f"Summarize the following article:\n\n{sample['article']}\n\nSummary:"
        
        # Tokenize prompt only
        tokenized = self.tokenizer(
            prompt,
            max_length=self.config.get("max_prompt_length", 512),
            truncation=True,
            padding=False,  # TRL handles padding
            return_tensors=None
        )
        
        ppo_samples.append({
            "input_ids": tokenized["input_ids"],
            "attention_mask": tokenized["attention_mask"],
            "query": prompt,  # For reward computation
            "reference": sample["summary"],  # For reward computation
            "article": sample["article"],  # Preserve for reward function
            "id": sample["id"]  # Preserve for tracking
        })
    
    return Dataset.from_list(ppo_samples)
```

#### 3. Update reward integration
```python
def compute_batch_rewards(self, prompts: List[str], summaries: List[str]) -> List[float]:
    """Integrate with TRL's reward computation cycle."""
    rewards = []
    for prompt, summary in zip(prompts, summaries):
        # Extract article from standardized prompt
        article = self._extract_article_from_prompt(prompt)
        reward = self.reward_function(article, summary)
        rewards.append(reward)
    return rewards
```

### Core Implementation
- [ ] `src/rlvr_summary/training/ppo_trainer.py` - Main integration changes
- [ ] `src/rlvr_summary/data/loaders.py` - Add tokenization support (optional)
- [ ] `configs/training/ppo_baseline.yaml` - Update config for new format

### Documentation Updates
- [ ] `src/rlvr_summary/data/README.md` - Document TRL format integration
- [ ] `REWARD_SYSTEM.md` - Update TRL integration section
- [ ] `README.md` - Update training pipeline documentation

### Test Updates
- [ ] `tests/test_training.py` - Update dataset creation tests
- [ ] `tests/test_data_pipeline.py` - Add TRL format conversion tests
- [ ] Create `tests/test_trl_integration.py` - New integration tests

### Demo Updates  
- [ ] `demo_training_implementation.py` - Update demo to show TRL format
- [ ] `scripts/train.py` - Ensure compatibility with new format
- [ ] `scripts/train_simple.py` - Update for TRL integration

### Functionality
- [ ] **Dataset conversion**: Raw data → TRL format works correctly
- [ ] **Training integration**: PPOTrainer accepts converted datasets
- [ ] **Reward computation**: Custom rewards integrate with TRL cycle
- [ ] **Memory efficiency**: No memory leaks or excessive usage
- [ ] **Performance**: Training speed matches or exceeds custom loop

### Compatibility
- [ ] **Backward compatibility**: Existing configs still work
- [ ] **Error handling**: Clear error messages for format issues
- [ ] **Logging**: W&B integration continues working
- [ ] **Checkpointing**: Model saving/loading works correctly

### Testing
- [ ] **Unit tests**: All existing tests pass with new format
- [ ] **Integration tests**: End-to-end training pipeline works
- [ ] **Performance tests**: Memory and speed benchmarks
- [ ] **Demo scripts**: All demo scripts work with new format

### Phase 1: Preparation
1. Create TRL format conversion utilities
2. Update tests to support both formats
3. Add configuration options for format selection

### Phase 2: Implementation  
1. Implement TRL dataset conversion
2. Update PPOTrainer integration
3. Test reward computation integration

### Phase 3: Validation
1. Run comprehensive testing
2. Performance benchmarking
3. Update documentation

### Phase 4: Cleanup
1. Remove custom training loop fallback
2. Clean up deprecated code
3. Final documentation updates

## Acceptance Criteria

- [ ] Data pipeline produces TRL-compatible Dataset objects
- [ ] PPOTrainer successfully trains with converted datasets
- [ ] Reward system integrates correctly with TRL training cycle
- [ ] All existing functionality preserved (evaluation, logging, checkpointing)
- [ ] Performance is equal to or better than custom implementation
- [ ] Documentation reflects new TRL integration
- [ ] All tests pass with new format
- [ ] Demo scripts demonstrate TRL integration
